src/Features/Core/Portable/Completion/Providers/AbstractKeywordCompletionProvider.cs

```diff
@@ -56,58 +43,57 @@ public override async Task ProvideCompletionsAsync(CompletionContext context)
-                }
             }
         }
 
-        private async Task<IEnumerable<CompletionItem>> RecommendCompletionItemsAsync(Document document, int position, CancellationToken cancellationToken)
+        private async Task<ImmutableArray<CompletionItem>> RecommendCompletionItemsAsync(Document document, int position, CancellationToken cancellationToken)
         {
+            var syntaxContextService = document.GetRequiredLanguageService<ISyntaxContextService>();
             var semanticModel = await document.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
-            var contextService = document.GetRequiredLanguageService<ISyntaxContextService>();
-            var syntaxContext = (TContext)await contextService.CreateContextAsync(document.Project.Solution.Workspace, semanticModel, position, cancellationToken).ConfigureAwait(false);
+            var syntaxContext = (TContext)await syntaxContextService.CreateContextAsync(document.Project.Solution.Workspace, semanticModel, position, cancellationToken).ConfigureAwait(false);
             var keywords = await RecommendKeywordsAsync(document, position, syntaxContext, cancellationToken).ConfigureAwait(false);
-            return keywords?.Select(k => CreateItem(k, syntaxContext));
+            return keywords.SelectAsArray(k => CreateItem(k, syntaxContext, cancellationToken));
         }
 
-        protected static ImmutableArray<string> s_Tags = ImmutableArray.Create(WellKnownTags.Intrinsic);
-
-        protected static CompletionItemRules s_keywordRules = CompletionItemRules.Default;
-
-        protected abstract CompletionItem CreateItem(RecommendedKeyword keyword, TContext context);
-
-        private async Task<IEnumerable<RecommendedKeyword>> RecommendKeywordsAsync(
+        private async Task<ImmutableArray<RecommendedKeyword>> RecommendKeywordsAsync(
             Document document,
             int position,
             TContext context,
             CancellationToken cancellationToken)
         {
             var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
-            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
             if (syntaxFacts.IsInNonUserCode(syntaxTree, position, cancellationToken))
-            {
-                return null;
-            }
+                return ImmutableArray<RecommendedKeyword>.Empty;
 
-            var set = new HashSet<RecommendedKeyword>();
+            using var _1 = ArrayBuilder<Task<ImmutableArray<RecommendedKeyword>>>.GetInstance(out var tasks);
             foreach (var recommender in _keywordRecommenders)
+                tasks.Add(Task.Run(() => recommender.RecommendKeywordsAsync(position, context, cancellationToken), cancellationToken));
+
+            await Task.WhenAll(tasks).ConfigureAwait(false);
```

Comment:
❔ Do we have any indication that sequential execution here is problematic? I would expect most calls to `RecommendKeywordsAsync` to complete synchronously, which would make the `Task.WhenAll` unnecessary and potentially expensive.

Reply:
The goal would be to have the 100ish C# providers (or 200+ VB providers) just run as quickly as possible on all cores. I could do this as a parallel foreach i suppose, but this seemed appropriately idiomatic.

Reply:
❗ Let's move the parallelism to a different PR. We can discuss it, but this PR is too big to deal with a change like that.

Reply:
can do.